integrations/langchain-js/src/BraintrustCallbackHandler.ts

```diff
@@ -0,0 +1,617 @@
+    top_p: pick(invocationParams?.top_p, invocationParams?.topP),
+    top_k: pick(invocationParams?.top_k, invocationParams?.topK),
+    max_tokens: pick(
+      invocationParams?.max_tokens,
+      invocationParams?.maxOutputTokens,
+    ),
+    frequency_penalty: invocationParams?.frequency_penalty,
+    presence_penalty: invocationParams?.presence_penalty,
+    response_format: invocationParams?.response_format,
+    tool_choice: invocationParams?.tool_choice,
+    function_call: invocationParams?.function_call,
+    n: invocationParams?.n,
+    stop: pick(invocationParams?.stop, invocationParams?.stop_sequence),
+  });
+
+  // Failsafe let's provide the invocation params as is
+  return !Object.keys(args).length ? invocationParams : args;
+};
+
+const pick = (...values: unknown[]) =>
+  values.find((value) => value !== undefined && value !== null);
+
+const outputFromGenerations = (
+  generations: Generation[][] | ChatGeneration[],
+) => {
+  const parsed = generations.flatMap((batch) => {
+    return Array.isArray(batch)
+      ? batch.map(parseGeneration)
+      : parseGeneration(batch);
+  });
+
+  return parsed;
+};
+
+const parseGeneration = (generation: Generation | ChatGeneration) => {
+  if ("message" in generation) {
+    return getMessageContent(generation.message);
+  }
+
+  if (generation.text) {
+    return generation.text;
+  }
+
+  // give up!
+};
+
+const inputFromMessages = (messages: BaseMessage[][]) => {
+  const parsed = messages.flatMap((batch) => batch.map(getMessageContent));
+  return parsed;
+};
+
+const getMessageContent = (message: BaseMessage) => {
+  let role = message.name ?? message.getType();
+
+  if (message.getType() === "human") {
+    role = "user";
+  } else if (message.getType() === "ai") {
+    role = "assistant";
+  } else if (message.getType() === "system") {
+    role = "system";
+  }
+
+  return cleanObject({
+    content: message.content,
+    role,
+    // @ts-expect-error Message may be any BaseMessage concrete implementation
+    tool_calls: message.tool_calls,
+    // @ts-expect-error Message may be any ToolMessage
+    status: message.status,
+    // @ts-expect-error Message may be any ToolMessage
+    artifact: message.artifact,
+  });
+};
+
+const cleanObject = (obj: Record<string, unknown>) =>
+  Object.fromEntries(
+    Object.entries(obj).filter(([key, value]) => {
+      if (value === undefined || value === null) return false;
+      if (Array.isArray(value) && value.length === 0) return false;
+      if (isObject(value) && Object.keys(value).length === 0) {
+        return false;
+      }
+      return true;
+    }),
+  );
+
+const safeParseSerializedJson = (input: string) => {
+  try {
+    return JSON.parse(input);
+  } catch {
+    return input;
+  }
+};
+
+const outputFromToolOutput = (output: unknown | ToolMessage) =>
+  output instanceof ToolMessage ? getMessageContent(output) : undefined;
+
+const outputFromChainValues = (output: unknown) => {
```

Comment:
ChainValues were the most complicated bit. I'm still not satisfied, but with a fallback to output at least users will not lose their data.